### PR TITLE
AP_OSD: apply aspect ratio correction to horizon

### PR DESCRIPTION
**Pull request by IamPete1**
Based on https://github.com/ArduPilot/ardupilot/issues/14853 from @FasaArg

I do not have a setup to test this. Happy to do test builds for others.

### CODE REVIEW

**Comment by FasaArg** [comment]
Thanks a lot for this @IamPete1 ,can your compile it for a Matek F765-wing? i could try it in both formats. I have a runcam micro eagle and i could switch between pal and ntsc in the same test rig.

**Comment by IamPete1** [comment]
[MatekF765-wing OSD angle test.zip](https://github.com/ArduPilot/ardupilot/files/4948737/MatekF765-wing.OSD.angle.test.zip)

@FasaArg I have attached both copter and plane. Thanks for testing.

**Comment by FasaArg** [comment]
Thanks a lot @IamPete1 , I tested the plane firmware (i should said this to you before, sorry) and seems to be working ok (to me at least)
Here are two videos from the DVR in PAL and NTSC to show the result beyond my initial appreciations.

NTSC: https://youtu.be/l3S_5oEpHDc

PAL: https://youtu.be/zJEBd0zp2KA

Let me know if you need further testing.
Thanks again.

**Comment by IamPete1** [comment]
@FasaArg thanks!, can you do a vid of what it is like at the moment on master? I have no feel for how much difference it would make.

https://firmware.ardupilot.org/Plane/latest/MatekF765-Wing/

**Comment by FasaArg** [comment]
Sure @IamPete1  here you have: https://youtu.be/5IKr6CykXHo

As you can see in 0º and 90º its the same , but the gradient between 0 and 90  shows the drift, is almost like a gopro superview effect. (on the horizon)

Of course is more notorious in NTSC because it have less horizontal lines.

image

**Comment by IamPete1** [comment]
I'm surprised is such a clear improvement! Very noticeable, Thanks

**Comment by FasaArg** [comment]
Thanks to you @IamPete1 I would never be able to code all you did regarding to auto select different coefficients depending on the video standar. Many thanks.

**Comment by IamPete1** [comment]
image
image
A before and after pic from @vierfuffzig

File: libraries/AP_OSD/AP_OSD_Backend.h [hunk 1]
```diff
@@ -42,6 +42,9 @@ class AP_OSD_Backend
     //update screen
     virtual void flush() = 0;
 
+    // return a correction factor used to display angles correctly
+    virtual float get_aspect_ratio_correction() {return 1;};
```
Comment (tridge) [inline libraries/AP_OSD/AP_OSD_Backend.h:46]: extra ;
should be const

Comment (IamPete1) [reply libraries/AP_OSD/AP_OSD_Backend.h]: fixed

**Pull request merged**